File: algorithms-data-structures/ts/src/algorithms/mixed/group-anagrams/group.ts
export function groupAnagrams(strs: string[]): string[][] {
  let res: string[][] = [];
  let soretedStrs = Array.from(strs.map((x) => [...x].sort().join("")));

  let store: Record<string, string[]> = {};
  for (let [i, word] of soretedStrs.entries()) {
    if (!store[word]) {
      store[word] = [strs[i]];
    } else {
      store[word].push(strs[i]);
    }
  }
  for (let [_, v] of Object.entries(store)) {
    res.push(v);
  }
  return res;
}

File: go/src/algorithms/mixed/group-anagrams/group.go
package groupanagrams

import "sort"

func groupAnagrams(strs []string) [][]string {
	var result [][]string
	store := make(map[string][]string)
	for i, word := range strs {
		sortedWord := sortString(word)
		store[sortedWord] = append(store[sortedWord], strs[i])
	}
	for _, list := range store {
		result = append(result, list)
	}
	return result
}

func sortString(s string) string {
	r := []rune(s)
	sort.Slice(r, func(i, j int) bool {
		return r[i] < r[j]
	})
	return string(r)
}
